### PR TITLE
GH-44607: [C++][Dev] Update bundled Thrift, update mirrors to use CDN

### LINKED ISSUE #44607: [C++] Thrift mirrors are outdated
Some observations:

- downloads.apache.org only has archives of Thrift 0.20.0 and 0.21.0 (current latest stable is 0.21.0)
- Apache appears to now (since 2021?) [have a CDN](https://apache.org/history/mirror-history.html): https://dlcdn.apache.org/thrift/. That also only has archives of 0.20.0 and 0.21.0.

https://apache.org/history/mirror-history.html gives more context and this sentence may explain why this list of URLs degraded over time:

> Any organization that has been hosting an Apache download mirror site, and that finds it useful for its own purposes, may continue to host the mirror site. 

So I think by continuing to use these mirror URLs we were living on borrowed time. The CDN is now the recommended way for users to download content from ASF projects.

So I think we actually need to do two things here:

- [x] Change the URL to just use the ASF CDN
- [x] Update our Thrift version in cpp/thirdparty/versions.txt to 0.20.0 or greater. Arrow C++ builds cleanly on my system when using Brew-provided Thrift 0.21.0 so it seems like this may be doable.

_Originally posted by @amoeba in https://github.com/apache/arrow/issues/44523#issuecomment-2450105564_

### PR DESCRIPTION
**Pull request by amoeba**
### Rationale for this change

Builds with bundled Thrift could fail because of permanently offline download mirror URLs. ASF now provides [a CDN](https://dlcdn.apache.org/).

### What changes are included in this PR?

1. Updates the pinned version for bundled Thrift from 0.16.0 to 0.20.0. We couldn't just update the URLs because the ASF CDN only hosts the last two versions of Thrift (0.20, 0.21.0). We didn't upgrade to Thrift 0.21.0 in this PR because it requires Boost 1.86.0 and we found incompatibilities with Boost 1.86.0 and `cpp/build-support/trim-boost.sh` which we use to publish a trimmed-down distribution of Boost for our build system to use.
2. Removes the set of URLs we were using to download Thrift in favor of just using the main CDN URL. We'll see how this works in practice.

### Are these changes tested?

Yes. I configured a build with `...-DThrift_SOURCE=BUNDLED...` and ran the full C++ test suite. I think further testing can be done in CI.

### Are there any user-facing changes?

No.

Closes https://github.com/apache/arrow/issues/44607
* GitHub Issue: #44607

### CODE REVIEW

**Comment by amoeba** [comment]
@github-actions crossbow submit test-r-linux-as-cran

**Comment by amoeba** [comment]
@github-actions crossbow submit -g cpp

**Comment by amoeba** [comment]
@github-actions crossbow submit -g cpp

**Comment by amoeba** [comment]
Windows MSVC is [failing with ](https://github.com/apache/arrow/actions/runs/11752225657/job/32743480413?pr=44685#step:9:1483):

> D:\a\arrow\arrow\build\cpp\thrift_ep-prefix\src\thrift_ep\lib\cpp\src\thrift\TUuid.cpp(22): fatal error C1083: Cannot open include file: 'boost/uuid/string_generator.hpp': No such file or directory

rhub/ubuntu-gcc-12 is [failing with](https://github.com/apache/arrow/actions/runs/11752225623/job/32743477204?pr=44685#step:6:1151):

> /tmp/Rtmpa7sH8K/file73956317ada/thrift_ep-prefix/src/thrift_ep/lib/cpp/src/thrift/TUuid.cpp:22:10: fatal error: boost/uuid/string_generator.hpp: No such file or directory
   22 | #include <boost/uuid/string_generator.hpp>
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

**Comment by amoeba** [comment]
The two failing builds were both bundling Boost and Thrift and I was able to reproduce the failure locally. From looking around, it looks like you have to upgrade to Boost 1.86.0 if you use Thrift 0.21.0. To test that hypothesis, I lowered Thrift down to 0.20.0 in https://github.com/apache/arrow/pull/44685/commits/4886ad7e12b1c008d12e909d913dacaae15b620b to test. I think we can get away with bumping Boost to 1.86.0 and Thrift to 0.21.0 in this PR so I'll try that once CI finishes.

**Comment by amoeba** [comment]
With Boost 1.86.0 and Thrift 0.21.0, the MSVC build fails with:

```
2024-11-09T17:43:07.4309013Z [550/608] Linking CXX shared library debug\parquet.dll
2024-11-09T17:43:07.4310713Z FAILED: debug/parquet.dll debug/parquet.lib 
2024-11-09T17:43:07.4323359Z C:\Windows\system32\cmd.exe /C "cd . && "C:\Program Files\CMake\bin\cmake.exe" -E vs_link_dll --intdir=src\parquet\CMakeFiles\parquet_shared.dir --rc=C:\PROGRA~2\WI3CF2~1\10\bin\100226~1.0\x64\rc.exe --mt=C:\PROGRA~2\WI3CF2~1\10\bin\100226~1.0\x64\mt.exe --manifests  -- C:\PROGRA~2\MICROS~2\2019\ENTERP~1\VC\Tools\MSVC\1429~1.301\bin\Hostx64\x64\link.exe /nologo src\parquet\CMakeFiles\parquet_shared.dir\Unity\unity_5_cxx.cxx.obj src\parquet\CMakeFiles\parquet_shared.dir\Unity\unity_4_cxx.cxx.obj src\parquet\CMakeFiles\parquet_shared.dir\Unity\unity_3_cxx.cxx.obj src\parquet\CMakeFiles\parquet_shared.dir\Unity\unity_2_cxx.cxx.obj src\parquet\CMakeFiles\parquet_shared.dir\Unity\unity_1_cxx.cxx.obj src\parquet\CMakeFiles\parquet_shared.dir\Unity\unity_0_cxx.cxx.obj src\parquet\CMakeFiles\parquet_shared.dir\__\generated\parquet_constants.cpp.obj src\parquet\CMakeFiles\parquet_shared.dir\__\generated\parquet_types.cpp.obj src\parquet\CMakeFiles\parquet_shared.dir\level_comparison_avx2.cc.obj src\parquet\CMakeFiles\parquet_shared.dir\level_conversion_bmi2.cc.obj  /out:debug\parquet.dll /implib:debug\parquet.lib /pdb:debug\parquet.pdb /dll /version:1900.0 /machine:x64  /NODEFAULTLIB:LIBCMT /debug /INCREMENTAL -Wl,--version-script=D:/a/arrow/arrow/cpp/src/parquet/symbols.map  debug\arrow.lib  "C:\Program Files\OpenSSL\lib\VC\libcrypto64MDd.lib"  "C:\Program Files\OpenSSL\lib\VC\libssl64MDd.lib"  thrift_ep-install\bin\thriftmd.lib  "C:\Program Files\OpenSSL\lib\VC\libcrypto64MDd.lib"  kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib && cd ."
2024-11-09T17:43:07.4347738Z LINK Pass 1: command "C:\PROGRA~2\MICROS~2\2019\ENTERP~1\VC\Tools\MSVC\1429~1.301\bin\Hostx64\x64\link.exe /nologo src\parquet\CMakeFiles\parquet_shared.dir\Unity\unity_5_cxx.cxx.obj src\parquet\CMakeFiles\parquet_shared.dir\Unity\unity_4_cxx.cxx.obj src\parquet\CMakeFiles\parquet_shared.dir\Unity\unity_3_cxx.cxx.obj src\parquet\CMakeFiles\parquet_shared.dir\Unity\unity_2_cxx.cxx.obj src\parquet\CMakeFiles\parquet_shared.dir\Unity\unity_1_cxx.cxx.obj src\parquet\CMakeFiles\parquet_shared.dir\Unity\unity_0_cxx.cxx.obj src\parquet\CMakeFiles\parquet_shared.dir\__\generated\parquet_constants.cpp.obj src\parquet\CMakeFiles\parquet_shared.dir\__\generated\parquet_types.cpp.obj src\parquet\CMakeFiles\parquet_shared.dir\level_comparison_avx2.cc.obj src\parquet\CMakeFiles\parquet_shared.dir\level_conversion_bmi2.cc.obj /out:debug\parquet.dll /implib:debug\parquet.lib /pdb:debug\parquet.pdb /dll /version:1900.0 /machine:x64 /NODEFAULTLIB:LIBCMT /debug /INCREMENTAL -Wl,--version-script=D:/a/arrow/arrow/cpp/src/parquet/symbols.map debug\arrow.lib C:\Program Files\OpenSSL\lib\VC\libcrypto64MDd.lib C:\Program Files\OpenSSL\lib\VC\libssl64MDd.lib thrift_ep-install\bin\thriftmd.lib C:\Program Files\OpenSSL\lib\VC\libcrypto64MDd.lib kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib /MANIFEST /MANIFESTFILE:src\parquet\CMakeFiles\parquet_shared.dir/intermediate.manifest src\parquet\CMakeFiles\parquet_shared.dir/manifest.res" failed (exit code 1104) with the following output:
2024-11-09T17:43:07.4362016Z LINK : warning LNK4044: unrecognized option '/Wl,--version-script=D:/a/arrow/arrow/cpp/src/parquet/symbols.map'; ignored
2024-11-09T17:43:07.4363542Z LINK : fatal error LNK1104: cannot open file 'thrift_ep-install\bin\thriftmd.lib'
```

**Comment by amoeba** [comment]
It looks like our build may be conflicting with a change in Thrift that appeared in 0.21.0:

https://github.com/apache/thrift/blob/b65ec607db09f2efdb7397acde88fd53cfb97f6b/build/cmake/DefineInstallationPaths.cmake#L23-L29

```
# For MSVC builds, install shared libs to bin/, while keeping the install
# dir for static libs as lib/.
if(MSVC AND BUILD_SHARED_LIBS)
    set(LIB_INSTALL_DIR "bin${LIB_SUFFIX}" CACHE PATH "The library install dir (default: bin${LIB_SUFFIX})")
else()
    set(LIB_INSTALL_DIR "lib${LIB_SUFFIX}" CACHE PATH "The library install dir (default: lib${LIB_SUFFIX})")
endif()
```

**Comment by amoeba** [comment]
@github-actions crossbow submit -g cpp

**Comment by amoeba** [comment]
@assignUser this is ready for a review. Three of the four failed crossbow jobs looks like test flakiness but another set of eyes would be good. The `test-debian-12-cpp-amd64` fails due to a related issue but for ORC. I think I could do a follow-up PR to update those URLs after this one.

File: cpp/thirdparty/versions.txt
Comment (assignUser) [inline cpp/thirdparty/versions.txt:144]: We use a trimmed down version of boost (10mb vs 142mb) to speed up downloads.[ Could you create the new archive ](https://github.com/apache/arrow/blob/03c3f8ed41573b611c4c6dfaf1ec166a9b78decb/cpp/build-support/trim-boost.sh#L23)and upload it to the artifactory (as a committer you can do that) and use that version here?

Comment (amoeba) [reply cpp/thirdparty/versions.txt:144]: Will do. I just tried to run the script and am getting failures on the bootstrap/b2 step. I'll take a look later today.

Comment (amoeba) [reply cpp/thirdparty/versions.txt:144]: I didn't have any luck and it appears to be caused by Boost 1.86.0 releasing with a broken `bcp`. See comment and really the whole thread at https://github.com/boostorg/bcp/pull/18#issuecomment-2311880441.

I think we could probably update the `trim-boost` script to work around this but we could also just switch back to Thrift 0.20.0 and Boost 1.81.0 until Boost 1.87.0.

**Comment by amoeba** [comment]
Due to the bcp issue with boost 1.86.0, I bumped Thrift down to 0.20.0 and Boost back to 1.81.0. We can update to Thrift 0.21.0 once  Boost 1.87.0 is out (and we can confirm the included bcp works).

**Comment by amoeba** [comment]
CI passed and PR description updated. @assignUser I think we can merge this without re-running all of the crossbow jobs and just be mindful during the 18.1.0 release process to look out for any issues.

File: cpp/cmake_modules/ThirdpartyToolchain.cmake
Comment (assignUser) [inline cpp/cmake_modules/ThirdpartyToolchain.cmake:1766]: Perfect!

**Review by assignUser — APPROVED** [review]
Agreed 👍 feel free to merge

**Comment by amoeba** [comment]
Merged, thanks.

**Pull request merged**
